Week1

EshauHuang opened the pull request:
你好，

我想請問我在執行 pull requests 時會顯示 "Can’t automatically merge. Don’t worry, you can still create the pull request."，

![Imgur](https://i.imgur.com/RU4tqgQ.png)


想請問該如何解決，謝謝。

EshauHuang committed:
hw1_v1

EshauHuang committed:
hw1_fin

EshauHuang committed:
hw3_fin

EshauHuang committed:
hw4_fin

EshauHuang committed:
hw5_fin

EshauHuang committed:
challenge_fin

zoeaeen13 commented:
我看了一下，你在 master 上面有 hw1 相關的 commit，這應該是跟 week1 分支上的 hw1 檔案衝突到，所以才出現「沒有辦法直接將 week1 合併到 master 」警示（你之前是在 master 分支改動而且推上遠端了），這裡我先 close 掉 PR。

---
給個提示，你先在 week1 分支拉遠端的 master 下來解衝突，再推一次和交作業

有任何問題都可以隨時問我～

zoeaeen13 commented on homeworks/week1/hw1.md:
```diff
@@ -1,2 +1,34 @@
+
+2. 確認是否所有檔案都已經 add且 commit
+
+![Imgur](https://i.imgur.com/35hqER3.png)
```
能從 github 下來的東西都是 commit 過的，此句不太正確。

這裡是因為一般拉下來後，我們習慣先以 `git status` 確認狀態，有時候拉新的 code 會有衝突需要解決，不過第一次拉專案是不會有這個問題的

zoeaeen13 commented on homeworks/week1/hw1.md:
```diff
@@ -1,2 +1,34 @@
+
+![Imgur](https://i.imgur.com/BxKdd9R.png)
+
+6. 將修改後的 hw1, commit 並 push 至 GitHub 
```
有附圖很用心喔～

不過這邊書寫順序怪怪的，應該是開 `week1` 分支，把所有作業寫好再一起推上 github，真的需要改才會再 commit 和 push，不過我猜你是要寫這個「繳交作業流程」才自己先截圖做測試

zoeaeen13 commented on homeworks/week1/hw3.md:
```diff
@@ -1,2 +1,26 @@
+
+![Imgur](https://i.imgur.com/2nf4F4m.png)
+
+2. 當我們需要變更檔案或是想要在某個位置創建檔案時，我們必須移動至當前位置才能達成目的。 
+   請輸入 cd 想要移動到的位置，例如:我目前的位置為 /c/users/Administrator ，而我想移動到 f 槽的資料夾，因此我輸入 cd  f:/eshau/test ，這樣我就切到了 f 槽的 test 資料夾了。
+
+![Imgur](https://i.imgur.com/3T3kjqe.png)
+
+3. 接下來我們輸入 mkdir 資料夾名稱，來創建自己想要的資料夾，之後再輸入 l(L的小寫)s 來確認是否建立成功。
+
+![Imgur](https://i.imgur.com/qZiyG8M.png)
+
+4. 如同第2點所說，我們必須移動到 wifi 資料夾內才可以在裡面新增檔案，因此輸入 cd wifi，然後輸入 touch 檔案名稱.副檔名，來創建檔案，之後再用 ls 來確認當前目錄下是否包含檔案。
```
流程清楚，建議在 markdown 格式下可以用 ` 來包住指令

比如 cd wifi 寫成 `cd wifi`，可讀性會更好

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -1,2 +1,49 @@
+
+* 假設你的笑話存放在 test 資料夾，這時我們必須先輸入 `git init` 來建立版本控制所需的檔案 `.git`，並輸入 `git status` 確認是否建立成功。
+
+*  Untracked files 下的紅色文字為你還，綠字則反之。如果要將 joke2 加入可以輸入 ` add joke2.txt `
```
「為你還」這邊缺漏字？😅，要加入「什麼」也要寫清楚，然後應該不是只用 `add` 指令

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -1,2 +1,49 @@
+
+![Imgur](https://i.imgur.com/FWqkAmQ.png)
+
+* 或是一次加入所有檔案，請輸入 `add .`。
```
附圖沒問題，但描述指令錯誤 `git add .`

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -1,2 +1,49 @@
+
+![Imgur](https://i.imgur.com/OrnQgKb.png)
+
+* 如有不想加入版本控制的檔案，請在資料夾內新建一個 `.gitignore` 檔案，建立方法為輸入`touch.gitignore`。
+* 接下來我們要在 Git Bash 中修改 `.gitignore` 文件，輸入 `vi(或是 vim) .gitignore` 來啟動文字編輯器。在此模式中按 "i "可以啟動編輯模式， "esc" 可以結束編輯模式，想要離開文字編輯器請先「離開編輯模式」後再輸入 `：q`。而我們要新增的是「不想要加入版本控制」的檔案的名稱，並在結束編輯模式後輸入 `:wq` 存檔並離開該模式。
+
+![Imgur](https://i.imgur.com/YpnpNS3.png)
+
+* 可以看到本來需要加入版本控制的檔案消失了，之後有想避開版本控制的文件都可以直接用此方法。
+
+![Imgur](https://i.imgur.com/5PmEh8r.png)
+
+![Imgur](https://i.imgur.com/O18edPV.png)
```
這邊寫的很棒，流程清楚，指令作用也有帶到，是適合初學者學習的文章

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -1,2 +1,49 @@
+
+![Imgur](https://i.imgur.com/hk6Lccl.png)
+
+* 如果之後修改了某些檔案，想建立一個新的版本我們可以直接輸入第3點的 `git commit -am "massage"` (如果是新增檔案的話就必須先 add 後才能 commit )，或是先 add 再輸入 `git commit -m "massage"`(前者是把所有檔案 add 後 commit；後者是只 commit 你 add 過的檔案，未 add 的檔案不會加入 commit，之後如果切換到該版本，沒有 add 的檔案會是上個版本沒有修改過的狀態)，都可以達到同樣的效果。
```
👍👍

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -1,2 +1,49 @@
+
+![Imgur](https://i.imgur.com/DYjJr3G.png)
+
+* 到這裡我們已經完成了初步的版本控制了，接著我們要學會如何自由切換版本，首先我們要先知道要切換版本的代號，輸入`git log`或是`git log --oneline`，我們可以看到 commit 後有一串代碼，之後只要輸入 `git checkout 代碼(或前7碼)`就可以做到版本切換，之後可以輸入 `git checkout master` 切換到最新版本。
```
`git checkout master` 不一定是切換到最新版本，master/main 只是 git/github 預設的初始分支，並不代表最新（這個之後工作時你會更能了解到）

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -1,2 +1,49 @@
+
+* 重新整理該網頁後，可以發現資料夾內的檔案都會顯示在上面，此時我們可以點擊畫面上的 commit 來知道每次更新版本新增或減少了什麼東西。
+
+* 當每次 commit 後我們都可以在 Git Bash 上用指令把本機的檔案同步到 GitHub 上，輸入`git push origin master` 就可以輕鬆同步；相反的，我們也可以在 GitHub 修改檔案並建立新的  commit，只要在修改完成後輸入 `git pull origin master` 就可以把 GitHub 做的修改同步到本機上了。
```
「當每次 commit 後我們都可以......」這句話太危險，切記不要每次 commit 完就把本機的檔案同步到 GitHub 上❗

push 要嚴謹一點，一旦推上去就沒辦法撤掉了，紀錄會永遠留在遠端、別人拉下同個專案也可以看到，應該是你做完一個功能或階段（比如完成 week1），全部確認沒問題了才會推上去

zoeaeen13 commented on homeworks/week1/hw5.md:
```diff
@@ -1,8 +1,140 @@
 
+前端就是呈現在使用者眼前的介面，例如網頁或是手機上的 APP 都是前端的一種。
+
+而後端就是指可以儲存、處理或傳送資料的伺服器或是資料庫，例如我們打了一串網址後，瀏覽器會透過網址的網域找到相對應的 DNS 伺服器 ，DNS 伺服器會給我們網站的 IP 位址，之後就可以請求 IP 位址的資料庫來取得該網站的資料，DNS 伺服器及 IP 位址的資料庫就是後端的一部份。
```
「給我們網站的 IP 位址，之後就可以請求 IP 位址的資料庫來取得該網站的資料」需要修正。

當我們從 DNS 得到 IP 位址之後，是將 request 送到這個 IP 位址對應的伺服器(Server)，Server 會視這個 request 的需要決定是否要跟它的資料庫溝通，**並非是前端去請求 IP 位址的資料庫**，然後 Server 跟資料庫溝通完拿到東西之後，會再處理成 response 返回到瀏覽器上。

我看你這題跟下面簡答題的敘述，顯然都是漏掉伺服器這一步，觀念上需要修正。
```
          DNS
       ↗↙ 
browser ⇄ Server ⇄ Database
```

zoeaeen13 commented on homeworks/week1/hw5.md:
```diff
@@ -1,8 +1,140 @@
 
 ## 假設我今天去 Google 首頁搜尋框打上：JavaScript 並且按下 Enter，請說出從這一刻開始到我看到搜尋結果為止發生在背後的事情。
 
+首先瀏覽器會先依照網址的網域(google.com)發送 request 到 DNS 來取得該網域的 IP 位址，之後我們可以從 Network 的 search?q=javascript...內的 Headers 發現我們使用 https(傳輸協定) 的 get 方法向 該 IP 位址(google 的資料庫)請求 JavaScript 搜尋結果的資料，而在 status code 中我們看到它回傳了 200，此數字代表請求成功了，因此可以看到之後對方開始傳送一些編碼(圖片、文字等等)，這些編碼經過瀏覽器的解析後將帶有圖片及文字的畫面呈現在我們眼前。
```
步驟需要修正，可以先參考其他同學的作業解釋

zoeaeen13 commented on homeworks/week1/挑戰題.md:
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+
+for ((i=1;i<=$1;i++))
+do 
+	touch "$i.js"	
+echo "檔案建立完成"	
+done
```
1. 空行太多囉，調整一下排版
2. 然後你可以直接用 num.sh 檔案放上來，不用寫成 markdown 文件
3. 現在執行的話，輸入幾次就會印幾次「檔案建立完成」，要想想看怎麼只讓它印一次提醒
---
進一步優化：如果使用者沒有輸入參數，就會跳 `./num.sh: line 1: ((: i<=: syntax error: operand expected (error token is "=")`，可以想看看怎麼做防呆、轉換成使用者看得懂的提醒文字

zoeaeen13 commented on homeworks/week1/hw5.md:
````diff
@@ -1,8 +1,140 @@
+   478c064..6b00e25  master     -> origin/master
+```
+
+但本機端的檔案目前還看不出變化，因此我們使用 Sourcetree 來觀察發生什麼事。
````
自己還開 Sourcetree 來做一遍嗎？真的很用心耶！

zoeaeen13 commented on homeworks/week1/hw5.md:
````diff
@@ -1,8 +1,140 @@
+
+[參考資料](https://zlargon.gitbooks.io/git-tutorial/content/branch/rebase.html)
+
+#### 2. 刪除 Commit
+
+Rebase 指令還可以輕鬆做到管理 Commit，如果我們想把 first 分支的 update 2 刪除的話要怎麼做呢?
+
+方法如下:
+
+```
+$ git rebase -i 7896494 //代表可以修改 add Rebase.txt 之後的 Commit 
+```
+
+![Imgur](https://i.imgur.com/3qNZr9t.png)
+
+這是一個 vim 編輯器，我們將 pick 修改成 drop 來達成刪除 Commit 的目的。
+
+![Imgur](https://i.imgur.com/B6da96Y.png)
+
+這樣我們就成功刪除 update 2 了!
````
看你的筆記我自己也是收穫很多！這些不太常用到就會忘記👍👍

zoeaeen13 reviewed (commented):
作業很棒！筆記和流程真的寫得認真，還自己一個個截圖配文太用心了，想必在撰寫過程中花費很大心力👍👍 少部分觀念有些可以修正的地方，及有缺字需調整，這裡我就先不 merge～

等你改完推上來，可以再 tag 我來看，看你的進度和筆記好詳細，自己也複習不少東西，如果還在上班真的是很厲害！佩服～

EshauHuang committed:
hw1_fix

因為作業上的圖片是以 hw1 為分支做的，因此沒有將 week1 修正成 hw1，這邊的 hw1 基本上代表我們正式繳交作業的 week1。

EshauHuang committed:
hw1_fix

EshauHuang committed:
hw3_fix

將作業內的指令都使用 ` 包覆，來增加可讀性。

EshauHuang committed:
fix_hw3

EshauHuang committed:
hw4_fix

自己看的時候都發現了一堆錯誤。之後檢查作業要在更仔細一點，會在 push 之前再順做最後一次總確認。

EshauHuang committed:
hw5_fix

修正觀念:
1. request  是經由瀏覽器 -> 作業系統(windows) -> 網卡，發送至 IP 位址的 Server
2. Sever 不是資料庫，而是與資料庫做溝通的一個伺服器

EshauHuang committed:
hw5_fix

EshauHuang committed:
add_num.sh

EshauHuang committed:
changeToNum

EshauHuang committed:
update num.sh

zoeaeen13 commented on homeworks/week1/num.sh:
```diff
@@ -1,8 +1,10 @@
-done
-echo "檔案建立完成"	
\ No newline at end of file
+if [ $1 -gt 0 ]; then
+	for ((i=1;i<=$1;i++)); do
+		touch "$i.js"
+	done
+echo "檔案建立完成"	
+else
+	echo "錯誤，請輸入大於 0 的數字"
+fi
```
修正讚讚，雖然沒輸入東西還是會先出現奇怪的那行，但已經有給使用者提醒了 👏
```
./num.sh: line 1: [: -gt: unary operator expected
錯誤，請輸入大於 0 的數字
```

zoeaeen13 commented on homeworks/week1/hw5.md:
```diff
@@ -2,11 +2,15 @@
 前端就是呈現在使用者眼前的介面，例如網頁或是手機上的 APP 都是前端的一種。
 
-而後端就是指可以儲存、處理或傳送資料的伺服器或是資料庫，例如我們打了一串網址後，瀏覽器會透過網址的網域找到相對應的 DNS 伺服器 ，DNS 伺服器會給我們網站的 IP 位址，之後就可以請求 IP 位址的資料庫來取得該網站的資料，DNS 伺服器及 IP 位址的資料庫就是後端的一部份。
+而後端就是指可以儲存、處理或傳送資料的伺服器或是資料庫，例如我們打了一串網址後，瀏覽器會發送一個 request 經由瀏覽器 -> 作業系統(windows) -> 網卡，透過網址的網域找到相對應的 DNS 伺服器，DNS 會傳給我們一個 IP 位址，經由此 IP 我們可以找到對應的 Server(伺服器)，而 Server 可以幫我們向資料庫溝通並取得我們想要的資料，而 DNS、Server、資料庫就是後端的一部份。
```
觀念 ok

zoeaeen13 commented on homeworks/week1/hw4.md:
```diff
@@ -45,5 +45,5 @@
 * 重新整理該網頁後，可以發現資料夾內的檔案都會顯示在上面，此時我們可以點擊畫面上的 commit 來知道每次更新版本新增或減少了什麼東西。
 
-* 當每次 commit 後我們都可以在 Git Bash 上用指令把本機的檔案同步到 GitHub 上，輸入`git push origin master` 就可以輕鬆同步；相反的，我們也可以在 GitHub 修改檔案並建立新的  commit，只要在修改完成後輸入 `git pull origin master` 就可以把 GitHub 做的修改同步到本機上了。
+* ~~當每次 commit 後我們都可以在 Git Bash 上用指令把本機的檔案同步到 GitHub 上~~ 切記不要每次 commit 後就 push 至遠端，最好是做到一個階段或做完一個功能並確認完全沒問題再上傳。輸入`git push origin master` 就可以同步至遠端；相反的，我們也可以在 GitHub 修改檔案並建立新的 commit，只要在修改完成後輸入 `git pull origin master` 就可以把 GitHub 做的修改同步到本機上了。
```
可是你還是每一個 commit 都 push 一遍 🤣 我的信箱有跳通知

zoeaeen13 reviewed (approved):
作業沒問題～恭喜完成第一週！改的很認真喔，go go 繼續加油
